plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Outline.hs

```diff
@@ -0,0 +1,131 @@
+    Just (toNormalizedFilePath' -> fp) -> do
+      mFields <- liftIO $ runIdeAction "cabal-plugin.fields" (shakeExtras ideState) (useWithStaleFast ParseCabalFields fp)
+      case fmap fst mFields of
+        Just fieldPositions -> pure $ LSP.InR (LSP.InL allSymbols)
+          where
+            allSymbols = mapMaybe documentSymbolForField fieldPositions
+        Nothing -> pure $ LSP.InL []
+    Nothing -> pure $ LSP.InL []
+
+documentSymbolForField :: Field Position -> Maybe LSP.DocumentSymbol
+documentSymbolForField (Field (Name pos fieldName) fieldLines) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeASCII fieldName,
+        LSP._kind = LSP.SymbolKind_Object,
+        LSP._children = Just $ mapMaybe documentSymbolForFieldLine fieldLines
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII fieldName
+documentSymbolForField (Section (Name pos fieldName) sectionArgs fields) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeASCII fieldName,
+        LSP._kind = LSP.SymbolKind_Object,
+        LSP._children =
+          Just
+            ( mapMaybe documentSymbolForField fields
+                ++ mapMaybe documentSymbolForSectionArgs sectionArgs
+            )
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII fieldName
+
+documentSymbolForSectionArgs :: SectionArg Position -> Maybe LSP.DocumentSymbol
+documentSymbolForSectionArgs (SecArgName pos identifier) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeLatin1 identifier,
+        LSP._kind = LSP.SymbolKind_Variable,
+        LSP._children = Nothing
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII identifier
+documentSymbolForSectionArgs (SecArgStr pos quotedString) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeLatin1 quotedString,
+        LSP._kind = LSP.SymbolKind_Constant,
+        LSP._children = Nothing
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII quotedString
+documentSymbolForSectionArgs (SecArgOther pos string) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeLatin1 string,
+        LSP._kind = LSP.SymbolKind_String,
+        LSP._children = Nothing
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII string
+
+documentSymbolForFieldLine :: FieldLine Position -> Maybe LSP.DocumentSymbol
+documentSymbolForFieldLine (FieldLine pos line) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeLatin1 line, -- since there is no ascii invariant (?)
+        LSP._kind = LSP.SymbolKind_Field,
+        LSP._children = Nothing -- can't delete even though the base case covers this (?)
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII line
+
+cabalPositionToLSPRange :: Position -> LSP.Range
```

Comment:
Worth noting that this just makes a single point range!

Reply:
I will document this as well.
The `readFields` returns a single position per field, so this choice of range seems reasonable for a default case.